closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -469,39 +471,50 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
+        // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+        byte[] keyCopy = key.clone();
+        byte[] ivCopy = iv.clone();
+
+        // need to synchronize this block as if there are concurrent calls to decryptFinal
+        // the shared variables aadBuffer and ibuffer may corrupt input and cause a segfault
+        // in openssl code
         if (len < 0) {
             throw new ProviderException("Input length is negative");
         }
-        if (len < tagLenBytes - ibuffer.size()) {
+
+        int bufSize = 0;
+        synchronized (this) {
+            bufSize = ibuffer.size();
+            if (len > 0) {
+                ibuffer.write(in, inOfs, len);
+            }
+
+            // refresh 'in' to all buffered-up bytes
+            in = ibuffer.toByteArray();
+            ibuffer.reset();
+        }
+
+        if (len < (tagLenBytes - bufSize)) {
             throw new AEADBadTagException("Input too short - need tag");
         }
-        if (len > MAX_BUF_SIZE - ibuffer.size()) {
+
+        if (len > (MAX_BUF_SIZE - bufSize)) {
             throw new ProviderException("SunJCE provider only supports "
                 + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
         }
 
-        if (out.length - outOfs < len  + ibuffer.size() - tagLenBytes) {
+        if (out.length - outOfs < len  + bufSize - tagLenBytes) {
```

Comment:
Please restore the parentheses from 7a564cc07d6534fcd88a11ef08dc84b0511064cd.

Reply:
Done